fem/qspace.cpp

```diff
@@ -35,6 +37,47 @@ void QuadratureSpaceBase::ConstructIntRules(int dim)
    }
 }
 
+void QuadratureSpaceBase::ConstructWeights() const
+{
+   // First get the Jacobian determinants (without the quadrature weight
+   // contributions).
+   weights = GetGeometricFactorWeights();
+
+   // Then scale by the quadrature weights.
+   const IntegrationRule &ir = GetIntRule(0);
+   const int N = size;
+   const int n = ir.Size();
+   double *d_weights = weights.ReadWrite();
+   const double *d_w = ir.GetWeights().Read();
+
+   mfem::forall(N, [=] MFEM_HOST_DEVICE (int i)
+   {
+      d_weights[i] *= d_w[i%n];
+   });
+}
+
+const Vector &QuadratureSpaceBase::GetWeights() const
+{
+   if (weights.Size() == 0) { ConstructWeights(); }
+   return weights;
+}
```

Comment:
This will return the wrong results if the Mesh nodes have been updated or the GeometricFactors have been updated, but the internal `QuadratureSpace` values have not been updated. The easiest check to ensure validity would be to compare the pointer of the GeometricFactor class used in the last calculation to current one stored on the mesh. If they pointers are equal then the weights are the same, but if they are different then you should need to reconstruct the weights: https://github.com/mfem/mfem/blob/master/mesh/mesh.hpp#L1730-1747

Reply:
I'd prefer to see this check added just as I do think this is an issue users will run into especially as there is no external way for them to update the weights themselves.

Reply:
Really good point, thanks for noticing that.

In my testing, comparing the pointer wasn't robust (sometimes the same memory address was used for the new geometric factors, so the weights weren't refreshed).

Instead I added `Mesh::nodes_sequence`, analogous to `Mesh::sequence`, to know when the cache has been invalidated (commits 6d35c65f954d11a8c7f66dd9c3baf361c11f0645 and 8d234b5358e9f2306116deee4866032aad5a233a).

Reply:
Thanks for adding the above fix!

Reply:
@tzanio, @v-dobrev, you might want to take a look at the addition of `Msh::nodes_sequence` to see if that makes sense. Related to #3343.

Reply:
I think the field `Mesh::nodes_sequence` should be added to `Mesh::Swap` and probably other methods of this type.

I noticed a call `GetIntRule(0)` in `QuadratureSpaceBase::ConstructWeights()` -- it will be good to guard that for empty local meshes.

Reply:
The empty meshes are handled in 4df15bf52605c3c84329e8b4718dcbda219ab51d.

I made some changes to handle `nodes_sequence` in d461ee1bca16d6c6817e581d7c48185fc2101893. I think it's OK, but not 100% sure I covered everything properly…